x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeSourceHandler.java

```diff
@@ -83,6 +84,11 @@ public ExchangeSourceHandler(int maxBufferSize, Executor fetchExecutor, ActionLi
         }));
     }
 
+    public void onFinishEarly(Runnable finishEarlyHandler) {
+        // TODO: not sure this is the best way but we need to know when the exchange source is finished early to set exec info
```

Comment:
Can we extend SubscribableListener in TransportEsqlQueryAction to allow overriding the response as partial before interrupting the result?

Reply:
I'm not sure it's possible as we don't have enough information at that point, but I'll double check.

Reply:
The problem here seems to be the listener does not know why it finished, and when we call finishEarly we don't have access to the `executionInfo` we need to modify - only to `ExchangeSourceHandler` which does not have it.

Reply:
Also, when we create `SubscribableListener` in execute(), we do not have `executionInfo` yet - we create it later in `innerExecute`. I could try to move its creation to earlier point but it may create other complications...

Reply:
Actually no, I am wrong, for async it's created in time, so it may be possible.